### DISCUSSION TITLE
Using non-ORM Postgres code

### DISCUSSION THREAD

**Discussion by troelsarvin** [diesel-rs/diesel#3512] [Q&A]
I ran across Diesel because the other Rust crates for Postgres seem to lack the ability to perform gssapi (Kerberos) connections. Diesel's ability to use all Postgres connectoin features is probably because Diesel is building on abstractions of the libpq C library whereas the other crates seem to try covering even the low-level protocol stuff themselves.

So I'm glad that I've run in to Diesel. However, my task involves raw SQL only, so I'm wondering if there is a way to opt out of ORM features? Has it been considered to separate the non-ORM code of Diesel's Postgres support into a separate crate (as an alternative to the "postgres" crate)?

**Comment by weiznich** [comment]
We do not provide such a feature flag as we just did not had any feature request to separate ORM code from "non-ORM" code. I would be interested in answers to the following questions to better understand why that might be required for certain use-cases:

* Which part of diesel do you consider ORM and which do you consider "non-ORM"?
* Why is it necessary to separate these parts into different crates/behind different feature flags opposed to "just" using one or the other part?

**Comment by troelsarvin** [comment]
With ORM I'm thinking about the query builder, and the migration mechanisms. About separating out: I could very well be wrong about the need for that, but my gut reaction was to think the ORM features incur "bloat" for those not using it?

**Reply by weiznich** [reply]
The migration mechanism is already not part of diesel itself, but in the `diesel_migrations` crate. For the query builder part: On the one hand it's hardly possible to move out the core parts of the query builder as these are also used by the raw sql functionality. On the other hand: The other parts only mostly at compile time as zero sized structs. That means the compiler will happily optimize anything away that's not used by your code. (And it will also provide an mostly zero cost abstraction for the parts that are used)

**Reply by troelsarvin** [reply]
I see what you mean. I guess I'm mainly looking for a way not to have to define a struct just to be able to extract data from a query.

So eg instead of
```
    struct Q {
        #[diesel(sql_type = Text)]
        res: String,
    }
    let q = sql_query("SELECT res FROM ... complicated stuff")
        .load::<Q>(conn)?;

```
if one could just do something alike

```
    let q = sql_query("SELECT res FROM ... complicated stuff")
        .load_singleton::<String>(conn)?;
```
(having other load-like expressions for rows, dicts, ...)

**Reply by weiznich** [reply]
With `sql_query` it is required to have that struct, because we somehow need to get the field name of the result field. That's not there if you just provide a target type.

That written: There are two other similar more low-level interfaces in diesel. These are not documented as well as `sql_query`, but depending on your use-case they might be more useful:

* [`diesel::dsl::sql`](https://docs.diesel.rs/2.0.x/diesel/dsl/fn.sql.html), which constructs a typed sql fragment. That can be a whole query, which means `diesel::dsl::sql::<Text>("SELECT res from …").load<String>(conn)?` just works. (Yes that's designed for extending the DSL in the first place, but it can be used to execute queries as well)
* [`LoadConnection::load`](https://docs.diesel.rs/2.0.x/diesel/connection/trait.LoadConnection.html#tymethod.load) which is arguable even more low level. It returns an iterator over the returned rows for anything that can be executed as query. That allows you to inspect the field types beforehand and provide your own matching. (That part is mostly undocumented as that's currently nothing many users do use. I would be open to add/accept more documentation there)

**Comment by troelsarvin** [comment]
Georg has provided useful answers, and I suggest the case be closed.
